pkg/securelaunch/helpers.go

```diff
@@ -0,0 +1,270 @@
+		}
+		devices := storage.PartitionsByFsUUID(storageBlkDevices, s[0]) // []BlockDev
+		for _, device := range devices {
+			Debug("device =%s with fsuuid=%s", device.Name, s[0])
+			deviceId = device.Name
+		}
+	}
+
+	devicePath := filepath.Join("/dev", deviceId) // assumes deviceId is sda, devicePath=/dev/sda
+	Debug("Attempting to mount %s", devicePath)
+	var dev *diskboot.Device
+	var err error
+	if rw_option {
+		dev, err = diskboot.FindDeviceRW(devicePath) // FindDevice fn mounts , w rw option, devicePath=/dev/sda.
+	} else {
+		dev, err = diskboot.FindDevice(devicePath) // FindDevice fn mounts devicePath=/dev/sda.
+	}
+	if err != nil {
+		return "", "", fmt.Errorf("Mount %v failed, err=%v", devicePath, err)
+	}
+
+	Debug("Mounted %s", devicePath)
+	fPath := filepath.Join(dev.MountPath, s[1]) // mountPath=/tmp/path/to/target/file if /dev/sda mounted on /tmp
+	return fPath, dev.MountPath, nil
+}
+
+/*
+ * ScanKernelCmdLine() ([]byte, error)
+ * format sl_policy=<block device identifier>:<path>
+ * e.g sda:/boot/securelaunch.policy
+ * e.g 4qccd342-12zr-4e99-9ze7-1234cb1234c4:/foo/securelaunch.policy
+ */
+func ScanKernelCmdLine() ([]byte, error) {
+
+	Debug("ScanKernelCmdLine: scanning kernel cmd line for *sl_policy* flag")
+	val, ok := cmdline.Flag("sl_policy")
+	if !ok {
+		log.Printf("ScanKernelCmdLine: sl_policy cmdline flag is not set")
+		return nil, errors.New("Flag Not Set")
+	}
+
+	// val is of type sda:path
+	mntFilePath, mountPath, e := GetMountedFilePath(val, false) // false means readonly mount
+	if e != nil {
+		log.Printf("ScanKernelCmdLine: GetMountedFilePath err=%v", e)
+		return nil, fmt.Errorf("ScanKernelCmdLine: GetMountedFilePath err=%v", e)
+	}
+	Debug("ScanKernelCmdLine: Reading file=%s", mntFilePath)
+
+	d, err := ioutil.ReadFile(mntFilePath)
+	if e := mount.Unmount(mountPath, true, false); e != nil {
+		log.Printf("Unmount failed. PANIC")
+		panic(e)
+	}
+
+	if err != nil {
+		// - TODO: should we check for end of file ?
+		log.Printf("Error reading policy file:mountPath=%s, passed=%s\n", mntFilePath, val)
+		return nil, fmt.Errorf("Error reading policy file:mountPath=%s, passed=%s\n", mntFilePath, val)
+	}
+	return d, nil
+}
+
+/*
+ *  scanBlockDevice(mountPath string) ([]byte, bool)
+ *	recursively scans an already mounted block device inside directories
+ *	"/", "/efi" and "/boot" for policy file
+ *
+ *	e.g: if you mount /dev/sda1 on /tmp/sda1,
+ *	then mountPath would be /tmp/sda1
+ *	and searchPath would be /tmp/sda1/, /tmp/sda1/efi, and /tmp/sda1/boot
+ *		respectively for each iteration of loop over SearchRoots slice.
+ */
+func ScanBlockDevice(mountPath string) ([]byte, bool) {
+
+	log.Printf("scanBlockDevice\n")
+	// scan for securelaunch.policy under /, /efi, or /boot
+	var SearchRoots = []string{"/", "/efi", "/boot"}
+	for _, c := range SearchRoots {
+
+		searchPath := filepath.Join(mountPath, c)
+		files := find.Find(context.Background(),
```

Comment:
optional: something to clean up resources here:

```go
ctx, cancel := context.WithCancel(context.Background())
defer cancel()
files := find.Find(ctx, ...
```

the cancel() will result in ctx.Done() returning something in this goroutine https://github.com/u-root/u-root/blob/f0f194ffe38f9dba3e24d8c1b604a9823c469d24/pkg/find/find.go#L181 which will then stop searching for files. (Otherwise, it keeps going while you don't even read from files anymore.)

Reply:
i like it.. will do